components/pluralrules/src/operands.rs

```diff
@@ -0,0 +1,275 @@
+            fraction_digits,
+        ) = if let Some(sep_idx) = abs_str.find('.') {
+            let int_str = &abs_str[..sep_idx];
+            let dec_str = &abs_str[(sep_idx + 1)..];
+
+            let integer_digits = u64::from_str(&int_str)?;
+
+            let dec_str_no_zeros = dec_str.trim_end_matches('0');
+
+            let num_fraction_digits0 = dec_str.len() as usize;
+            let num_fraction_digits = dec_str_no_zeros.len() as usize;
+
+            let fraction_digits0 = u64::from_str(&dec_str)?;
+            let fraction_digits =
+                if num_fraction_digits == 0 || num_fraction_digits == num_fraction_digits0 {
+                    fraction_digits0
+                } else {
+                    u64::from_str(&dec_str_no_zeros)?
+                };
+
+            let fraction: f64 = (fraction_digits as f64) / 10_f64.powi(num_fraction_digits as i32);
+
+            (
+                integer_digits as f64 + fraction,
+                integer_digits,
+                num_fraction_digits0,
+                num_fraction_digits,
+                fraction_digits0,
+                fraction_digits,
+            )
+        } else {
+            let integer_digits = u64::from_str(&abs_str)?;
+            (integer_digits as f64, integer_digits, 0, 0, 0, 0)
+        };
+
+        Ok(PluralOperands {
+            n: absolute_value,
+            i: integer_digits,
+            v: num_fraction_digits0,
+            w: num_fraction_digits,
+            f: fraction_digits0,
+            t: fraction_digits,
+        })
+    }
+}
+
+macro_rules! impl_integer_type {
+    ($ty:ident) => {
+        impl From<$ty> for PluralOperands {
+            fn from(input: $ty) -> Self {
+                PluralOperands {
+                    n: input as f64,
+                    i: input as u64,
+                    v: 0,
+                    w: 0,
+                    f: 0,
+                    t: 0,
+                }
+            }
+        }
+    };
+    ($($ty:ident)+) => {
+        $(impl_integer_type!($ty);)+
+    };
+}
+
+macro_rules! impl_signed_integer_type {
+    ($ty:ident) => {
+        impl TryFrom<$ty> for PluralOperands {
+            type Error = OperandsError;
+            fn try_from(input: $ty) -> Result<Self, Self::Error> {
+                let x = input.checked_abs().ok_or(OperandsError::Invalid)?;
+                Ok(PluralOperands {
+                    n: x as f64,
+                    i: x as u64,
+                    v: 0,
+                    w: 0,
+                    f: 0,
+                    t: 0,
+                })
+            }
+        }
+    };
+    ($($ty:ident)+) => {
+        $(impl_signed_integer_type!($ty);)+
+    };
+}
+
+// This algorithms allows us to use the Rust's decimal to float parser
+// using on-stack buffer.
+//
+// Once we start relying on Ryu, we should consider swapping it to use that.
+fn get_dec_str_from_f64(absolute_value: f64, buffer: &mut [u8]) -> Result<&str, IOError> {
```

Comment:
Shouldn't you say explicitly that the lifetime of the `&str` in result is tied to the lifetime of `buffer`?

Reply:
That's implicitly held, why would I explicitly add it? :)